benchmark/ycsb.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+Run benchmarks in various configurations for each defined datastore.
+"""
+
+import argparse
+import copy
+import logging
+import os
+import shutil
+import subprocess
+import time
+from dataclasses import dataclass
+from subprocess import Popen
+from typing import List
+
+import cimetrics.upload  # type: ignore
+import pandas as pd  # type: ignore
+
+import common
+from common import Store, wait_with_timeout, DESIRED_DURATION_S
+from stores import EtcdStore, LSKVStore
+
+logging.basicConfig(format="%(asctime)s %(levelname)s %(message)s", level=logging.INFO)
+
+BENCH_DIR = os.path.join(common.BENCH_DIR, "ycsb")
+
+
+# pylint: disable=too-many-instance-attributes
+@dataclass
+class YCSBConfig(common.Config):
+    """
+    Config holds the configuration options for a given benchmark run.
+    """
+
+    workload: str
+
+    def bench_name(self) -> str:
+        """
+        Get the name of the benchmark.
+        """
+        return "ycsb"
+
+
+class YCSBenchmark(common.Benchmark):
+    """
+    YCS benchmark.
+    """
+
+    def __init__(self, config: YCSBConfig):
+        self.config = config
+
+    def load_cmd(self, store: Store) -> List[str]:
+        """
+        Return the command to run the benchmark for the given store.
+        """
+        timings_file = os.path.join(self.config.output_dir(), "timings.csv")
+        bench = [
+            "bin/go-ycsb",
+            "load",
+            "etcd",
+            "--prop",
+            f"etcd.endpoints={self.config.scheme()}://127.0.0.1:{self.config.port}",
+            "--prop",
+            "measurementtype=raw",
+            "--prop",
+            f"measurement.raw.output_file={timings_file}",
+            "--property_file",
+            self.path_to_workload(),
+        ]
+        logging.debug("load cmd: %s", bench)
+        if self.config.tls:
+            bench += [
+                "--prop",
+                f"etcd.cacert_file={store.cacert()}",
+                "--prop",
+                f"etcd.cert_file={store.cert()}",
+                "--prop",
+                f"etcd.key_file={store.key()}",
+            ]
+        return bench
+
+    def run_cmd(self, store: Store) -> List[str]:
+        """
+        Return the command to run the benchmark for the given store.
+        """
+        timings_file = os.path.join(self.config.output_dir(), "timings.csv")
+        bench = [
+            "bin/go-ycsb",
+            "run",
+            "etcd",
+            "--prop",
+            f"etcd.endpoints={self.config.scheme()}://127.0.0.1:{self.config.port}",
+            "--prop",
+            "measurementtype=raw",
+            "--prop",
+            f"measurement.raw.output_file={timings_file}",
```

Comment:
go-ycsb doesn't currently support the raw output measurement style so will take a look at adding raising an issue on their repo.

Reply:
https://github.com/pingcap/go-ycsb/issues/255